### PR TITLE
prevent some cmake-lint warnings

### PR DESCRIPTION
**Pull request by ClausKlein**
export missing include path
cmake-format the files too
yamllint the CI config too

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,17 +1,16 @@
+---
```
Comment (friendlyanon) [inline .github/workflows/ci.yml:1]: GitHub Actions workflows don't support the entire YAML spec nor multidocument YAML files, so this is entirely superfluous.  
Please read the documentation about the workflow syntax: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions

Not even anchors and I wish that wasn't so.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,17 +1,16 @@
...
-      - cmake-3.12
-      - cmake-3.8
+      - develop
 
   pull_request:
     branches:
       - master
-      - cmake-3.12
-      - cmake-3.8
+      - develop
```
Comment (friendlyanon) [inline .github/workflows/ci.yml:13]: This repository doesn't have a `develop` branch. There is nothing to develop in fact. This repository is "done" as there is nothing meaningful to add to achieve its purpose. I didn't archive it only becuase CMake might further simplify some things in the future which I will integrate into this repository, but that boils down to just creating a new named branch for the current `master` (CMake 3.14) and adjusting a few lines.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -24,8 +23,8 @@ jobs:
         os: [macos, ubuntu, windows]
 
         include:
-          - { type: shared, build: YES }
-          - { type: static, build: NO }
+          - {type: shared, build: YES}
+          - {type: static, build: NO}
```
Comment (friendlyanon) [inline .github/workflows/ci.yml:27]: PR title is about cmake-lint, why is this file being touched anyway?

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,111 +1,104 @@
-cmake_minimum_required(VERSION 3.14)
+cmake_minimum_required(VERSION 3.14...3.20)
```
Comment (friendlyanon) [inline CMakeLists.txt:1]: Ranges only make sense if you make use of new policies conditionally, which is not the case for this repository, but you'd know that if you actually read the CML files.

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,111 +1,104 @@
...
 project(
-    example
-    VERSION 0.1.0
-    LANGUAGES CXX
+  example
+  VERSION 0.1.0
+  LANGUAGES CXX
 )
 
+option(BUILD_SHARED_LIBS "default" YES)
```
Comment (friendlyanon) [inline CMakeLists.txt:9]: `BUILD_SHARED_LIBS` must **never** be set in project CMLs.

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,111 +1,104 @@
...
+
 # ---- Warning guard ----
 
 # Protect dependents from this project's warnings if the guard isn't disabled
-set(example_warning_guard SYSTEM)
-if(example_INCLUDE_WITHOUT_SYSTEM)
-  set(example_warning_guard "")
-endif()
+## XXX why? CK
```
Comment (friendlyanon) [inline CMakeLists.txt:14]: Imported targets' (ones coming from `find_package`) include paths are implicitly treated as `SYSTEM`. If a user wants to vendor a library, without a warning guard they will have to deal with __your__ warnings in __his__ build tree.

To mirror behavior of imported targets and make users' life easier, this is necessary. If the user still wants warnings, they can be trivially toggled in multiple ways.

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,111 +1,104 @@
...
+##   set(example_warning_guard "")
+## endif()
 
 # ---- Declare library ----
 
 include(GenerateExportHeader)
 
-set(headers include/example/main.h)
+include(GNUInstallDirs)
+set(EXAMPLE_INCLUDE_DIRECTORY "${CMAKE_INSTALL_INCLUDEDIR}")
```
Comment (friendlyanon) [inline CMakeLists.txt:25]: Setting variables all over the place without structure (which are already provided by the deliniating `----` comments) just makes an unstructured, unreadable mess.

I advise strongly against using just `CMAKE_INSTALL_INCLUDEDIR` as the include path as that can and does cause issues in some cases. More details here: `https://github.com/Dobiasd/FunctionalPlus/issues/222`

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,111 +1,104 @@
...
+
+set(EXAMPLE_headers include/example/main.h)
 
-set(sources source/main.cpp)
+set(EXAMPLE_sources source/main.cpp)
```
Comment (friendlyanon) [inline CMakeLists.txt:29]: There is no `EXAMPLE` target, why the prefix? There is only a single target as well, so again, why the prefix?

File: CMakeLists.txt [hunk 9]
```diff
@@ -1,111 +1,104 @@
...
-    SOVERSION "${PROJECT_VERSION_MAJOR}"
-    EXPORT_NAME example
-    OUTPUT_NAME example
+  example_example
+  PROPERTIES CXX_VISIBILITY_PRESET hidden
+             VISIBILITY_INLINES_HIDDEN YES
+             VERSION "${PROJECT_VERSION}"
+             SOVERSION "${PROJECT_VERSION_MAJOR}"
+             EXPORT_NAME example
+             OUTPUT_NAME example
```
Comment (friendlyanon) [inline CMakeLists.txt:46]: This is a bogus formatting style, I will never understand support this insanity.

File: CMakeLists.txt [hunk 10]
```diff
@@ -1,111 +1,104 @@
...
 target_include_directories(
-    example_example
-    ${example_warning_guard}
-    PUBLIC
-    "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
-    "$<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/include>"
+  example_example # XXX ${example_warning_guard}
+  PUBLIC "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
+         "$<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/include>"
+         "$<INSTALL_INTERFACE:${EXAMPLE_INCLUDE_DIRECTORY}>"
```
Comment (friendlyanon) [inline CMakeLists.txt:53]: The interface include directories are already specified in the install rules. You also removed the warning guard. Unacceptable.

File: CMakeLists.txt [hunk 11]
```diff
@@ -1,111 +1,104 @@
...
 )
 
-target_compile_features(example_example PUBLIC cxx_std_14)
+target_compile_features(example_example PUBLIC cxx_std_17)
```
Comment (friendlyanon) [inline CMakeLists.txt:56]: What C++17 feature is in use by this library? While true that I arbitrarily chose C++14, there is nothing that would warrant raising the requirement.

File: CMakeLists.txt [hunk 12]
```diff
@@ -1,111 +1,104 @@
...
 )
 
-set(
-    example_INSTALL_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake/example"
+set(EXAMPLE_INSTALL_CMAKEDIR
+    "${CMAKE_INSTALL_LIBDIR}/cmake/example"
     CACHE STRING "CMake package config location relative to the install prefix"
 )
-mark_as_advanced(example_INSTALL_CMAKEDIR)
+mark_as_advanced(EXAMPLE_INSTALL_CMAKEDIR)
```
Comment (friendlyanon) [inline CMakeLists.txt:86]: Why change the casing? The GUI CMake clients will still group variables properly regardless of casing, these are not variables that go into configured header files for macro names and the `project` call was given `example` as a name not `EXAMPLE`.

File: CMakeLists.txt [hunk 13]
```diff
@@ -1,111 +1,104 @@
...
-    DESTINATION "${example_INSTALL_CMAKEDIR}"
-    COMPONENT example_Development
+  EXPORT exampleTargets
+  NAMESPACE example::
+  DESTINATION "${EXAMPLE_INSTALL_CMAKEDIR}"
+  COMPONENT example_Development
 )
 
-if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
```
Comment (friendlyanon) [inline CMakeLists.txt:102]: This is a pointless change. Please learn how the `if` command works: https://cmake.org/cmake/help/latest/command/if.html

File: all/CMakeLists.txt [hunk 14]
```diff
@@ -1,13 +1,14 @@
-cmake_minimum_required(VERSION 3.8)
+cmake_minimum_required(VERSION 3.14...3.20)
 
 project(exampleAll NONE)
 
-set(example_INCLUDE_WITHOUT_SYSTEM YES CACHE INTERNAL "")
+#XXX set(example_INCLUDE_WITHOUT_SYSTEM YES CACHE INTERNAL "")
```
Comment (friendlyanon) [inline all/CMakeLists.txt:5]: This is a superbuild CML for development. Why disable warnings during developments?

File: all/CMakeLists.txt [hunk 15]
```diff
@@ -1,13 +1,14 @@
...
 
 add_subdirectory(.. sub/root)
 
 # The above add_subdirectory call creates the necessary target(s), add a stub
 file(WRITE "${PROJECT_BINARY_DIR}/configs/exampleConfig.cmake" "")
 set(example_DIR "${PROJECT_BINARY_DIR}/configs" CACHE INTERNAL "")
 
+enable_testing()
```
Comment (friendlyanon) [inline all/CMakeLists.txt:13]: It's not a superbuild CML's responsibility to muck with testing. It's already enabled as necessary in the test's project CML.

File: test/CMakeLists.txt [hunk 16]
```diff
@@ -1,14 +1,14 @@
-cmake_minimum_required(VERSION 3.8)
+cmake_minimum_required(VERSION 3.14...3.20)
 
-project(example_tests CXX)
+project(example_tests LANGUAGES CXX)
```
Comment (friendlyanon) [inline test/CMakeLists.txt:3]: Unnecessary change. `project`'s short form takes languages after the project name. Please read the documentation: https://cmake.org/cmake/help/latest/command/project.html

**Review by friendlyanon — CHANGES_REQUESTED** [review]
If this is due to my review on the PR `https://github.com/chriskohlhoff/asio/pull/420`, then I have to say that it is quite sad that this is how you handle constructive input.

I would have preferred you asking questions if you didn't understand something. I'm willing to spend some extra time on trying to teach someone.

Aside from the bogus and wrong changes in this PR, your commit messages could also use some improvements. I can recommend reading https://chris.beams.io/posts/git-commit/ which tells you how to write meaningful and legible commit messages that are easy to make sense of.  
Of course there is https://www.conventionalcommits.org/ as well, but its use depends on the project and there are also conventions tied to issue trackers like Jira where applicable. Obviously those are not really the general solutions, so I recommend reading the first link and really internalizing it. Readable commit messages make review easier as well.

Another thing is that PRs should focus on one thing. Logically separate changes should go in separate PRs. Of course, when something as overarching as a build tool is the subject of a PR, it's quite inevitable that files with different concerns must be touched for meaningful changes. This wasn't the case here however, the CI workflow needed absolutely no changes.

**Comment by friendlyanon** [comment]
I'm going to go ahead and close this, as this is obviously someone not taking helpful and constructive feedback well.

Please refrain from such childlike behavior in the future.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ClausKlein) [reply .github/workflows/ci.yml:1]: may be, but prefer to use yamllint, and no warnings!

File: .github/workflows/ci.yml (hunk 3 above)
Comment (ClausKlein) [reply .github/workflows/ci.yml:27]: yamlint ...

File: CMakeLists.txt (hunk 6 above)
Comment (ClausKlein) [reply CMakeLists.txt:14]: I don't understand! it is not needed IMHO.

File: CMakeLists.txt (hunk 5 above)
Comment (ClausKlein) [reply CMakeLists.txt:9]: What do you mean with CML?
Were is this written?

File: CMakeLists.txt (hunk 8 above)
Comment (ClausKlein) [reply CMakeLists.txt:29]: cmake-lint warnings about lower case cmake variable names!

File: CMakeLists.txt (hunk 10 above)
Comment (ClausKlein) [reply CMakeLists.txt:53]: I know, anyway!

File: CMakeLists.txt (hunk 13 above)
Comment (ClausKlein) [reply CMakeLists.txt:102]: if(<variable|string> STREQUAL <variable|string>)
True if the given string or variable's value is lexicographically equal to the string or variable on the right.

I know, but everything in CMAKE is first a string!
is seems more clear to be explicit.

File: all/CMakeLists.txt (hunk 15 above)
Comment (ClausKlein) [reply all/CMakeLists.txt:13]: you don't understand what I does here!

I test the installed `cmake config package` 
this is now the `cmake master project `and need this!

Comment (friendlyanon) [reply all/CMakeLists.txt:13]: Please speak in a coherent way and don't resolve conversations. I will resolve them when I consider them resolved.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (friendlyanon) [reply .github/workflows/ci.yml:1]: Like I said, multidocument YAML files are not supported by GitHub Actions and this project doesn't use YAMLlint.

File: CMakeLists.txt (hunk 5 above)
Comment (friendlyanon) [reply CMakeLists.txt:9]: **CM**ake**L**ists.txt

File: CMakeLists.txt (hunk 6 above)
Comment (friendlyanon) [reply CMakeLists.txt:14]: With this warning guard dependents can use `add_subdirectory` (ie vendor a dependency) and still get the same behaviour as if they used `find_package`. It's a few lines of code for a very nice convenience function that can be toggled easily. I prefer a happy user to screwing him over with my warnings.

File: CMakeLists.txt (hunk 13 above)
Comment (friendlyanon) [reply CMakeLists.txt:102]: `if()`'s documentation is already explicit enough about what it does.
